core/lib/types/src/commitment.rs

```diff
@@ -209,14 +209,51 @@ impl L1BatchWithMetadata {
                         .to_vec(),
                 ),
                 Token::Bytes(self.metadata.l2_l1_messages_compressed.clone()),
-                Token::Bytes(self.header.pubdata_input.clone().unwrap()),
+                Token::Bytes(
+                    self.header
+                        .pubdata_input
+                        .clone()
+                        .unwrap_or(self.construct_pubdata()),
```

Comment:
Should be unwrap_or_else, because `construct_pubdata()` is called every time no matter if you need it or not.